### DISCUSSION TITLE
Prevent incorrect initialization

### DISCUSSION THREAD

**Discussion by me21** [bxparks/AceTime#120] [General]
Hello,

I found that if I create ZoneManager like this:

```c++
ace_time::CompleteZoneManager zoneManager(
  ace_time::zonedbc::kZoneRegistrySize, // note incorrect size passed, should be kZoneAndLinkRegistrySize
  ace_time::zonedbc::kZoneAndLinkRegistry,
  zoneProcessorCache);
```

the code hangs up while trying to enumerate them later:

```c++
  // here we pass the correct size
  zoneSorter.fillIndexes(indexes, ace_time::zonedbc::kZoneAndLinkRegistrySize);
  zoneSorter.sortIndexes(indexes, ace_time::zonedbc::kZoneAndLinkRegistrySize);

  // later...
  auto zone = zoneManager.getZoneForIndex(indexes[0]); // hangs up here
```

I understand it's technically not a bug within the library, but maybe there's a way to prevent incorrect size passing? The size of array `ace_time::zonedbc::kZoneAndLinkRegistry` is known at compile-time, so maybe it's possible to avoid passing its size directly?

**Comment by bxparks** [comment]
I've never been able to come up with a solution that is ergonomic and reliable. It always seems to depend on the programmer to do the right thing, because the C or C++ language does not provide any mechanism to bind the length information and pointer together in a robust way.

Yes the `ace_time::zonedbc::kZoneAndLinkRegistry` array is known at compile-time, but the constructor actually accepts a pointer, not a compile-time array. The constructor relies on the automatic decay of the array into a pointer. It is not guaranteed that we will always know the size of the array at compile time, because the constructor of the `ZoneManager` may happen in a context (e.g. inside a factory function) where we have only the pointer. We **always** have to rely on the programmer to pass in the correct length information as a separate parameter.

We could *suggest* that the programmer do something like:

```C++
ace_time::CompleteZoneManager zoneManager(
  sizeof(ace_time::zonedbc::kZoneAndLinkRegistry) / 
    sizeof(ace_time::zonedbc::kZoneAndLinkRegistry[0]),
  ace_time::zonedbc::kZoneAndLinkRegistry,
  zoneProcessorCache);
```

This only works if the registry is a compile-time array. It does not work if it's passed in as a pointer.

This also depends on the programmer to write the correct `sizeof()` expression, and that the programmer uses the correct `kZoneAndLinkRegistry` array inside that `sizeof()` expression. I've definitely been guilty of copy-and-pasting code like this, and forgetting to replace that array name, or length parameter. But when I write the expression from scratch, I get the `sizeof()` expression wrong a substantial fraction of the time. Using a `#define` macro may help, but then the programmer has to remember to either define the correct `#define` macro, or remember to do a `#include` of the correct header file that includes this `#define` macro, which can sometimes be a source of friction.

You could say that we should create a new class (let's call it a `SmartZoneRegistry` class) that binds the length and array together into a single object. But all we have done is push the problem one step back, because the code fragment that creates a new instance of the `SmartZoneRegistry` is the same bit of code that creates a `ZoneManager` class. The other problem with creating a new `SmartZoneRegistry` class is that it seems to create too much abstraction, for not much benefit. The AceTime library already has too many classes, and I've resisted creating helper classes unless absolutely necessary.

**Comment by me21** [comment]
I thought about changing `ZoneManager` constructor signature to accepting `std::vector` instead of pointer+size pair - would it make the library unavailable to 8-bit microcontrollers?.. Probably so.
Another thought was to make a template constructor which would accept `std::array` - hopefully its function calls can be inlined but it would add the size information.
These possibilities can be added to `ZoneManager` instead of replacing the existing constructor. Possibly the compiler can optimize the unused parts out, so the lib can still be used with 8-bit chips even though there would be a constructor accepting `std::vector`.

**Reply by bxparks** [reply]
I wrote in the other reply, my policy is to have no dependencies to the `std::` libraries for my Arduino libraries. The C++ libraries are too bloated, too big, and not universally available for the various Arduino-compatible platforms which are pegged to different versions of C++.

**Comment by me21** [comment]
Found this answer on StackOverflow: https://stackoverflow.com/a/60151658/2404492
What if `kZoneAndLinkRegistry` arrays would be hidden from the user of this library and `span`s provided instead?

I guess this is what you meant as `SmartZoneRegistry`, but these objects can be pre-created by the library, so that user doesn't have to write that code.

Sure, if the library user creates a custom registry, he would have to write the same code, but this change will reduce the number of users who can make a mistake, from all users of the library to only those who create their registries.

Just a thought.

**Reply by bxparks** [reply]
From the article, "If you are able to use C++20, you could use std::span". My Arduino libraries will never have any dependencies to the C++ standard libraries. Different Arduino-compatible platforms support different levels of C++ (C++11, C++14, C++17, I don't know if any of them has migrated to C++20 or 23 yet). Even when a platform claims a certain level of C++XX compatility, I'm not always convinced that the standard libraries are actually implemented well. Also, the C++ standard libraries tend to be overly complex and bloated, especially for embedded work.

Sure those `SmartZoneRegistry` instances could be pre-created. But I don't like the proliferation of extra helper classes. It makes the surface area of the API bigger, which means slightly harder to understand. It's a tradeoff... I tend to lean more towards smaller API, and fewer classes, rather than trying to mitigate all possible mistakes that a programmer can make.

I've personally found that custom registries is the common case, rather than using the predefined ones. New users may use the predefined ones. But the predefined registries tend to be too large for most microcontroller applications. So my expectation is that most people will use custom registries. For example, I normally use a registry of only 4-6 timezones. That slashes the memory consumption by 30-50 kB, which is a big deal for the ATmega328 or the lowerend STM32 processors with only 64kB or 128kB.

**Reply by me21** [reply]
The same answer has C++11 implementation of `span`, just in case. But yeah, if most people use custom registries, then this proposition is useless.

Another thought: would something like the following work, or is this syntax incorrect?..

```
template<size_t N>
ace_time::ZoneManagerTemplate(const ZI* zoneRegistry[N], ZoneProcessorCacheBaseTemplate<ZP>& zoneProcessorCache):
        mZoneRegistrar(N, zoneRegistry),
        mZoneProcessorCache(zoneProcessorCache)
    {}
```

e.g. add a templated constructor with array size as a template parameter.
